qa/tasks/ceph.py

```diff
@@ -1243,6 +1319,99 @@ def first_in_ceph_log(pattern, excludes):
+			     implemented for filestore")
+                    return
+
+    log.info("Using objectstore {}".format(objstore))
+    ceph_lvm = "sudo ceph-volume lvm "
+    for remote, roles_to_dev in remote_to_roles_to_devs.iteritems():
+        for osd, dev in roles_to_dev.iteritems():
+            log.info("OSD={}, dev={}".format(osd, dev))
+            cmd = ceph_lvm + "create --" + objstore + \
+                " --data "+ dev
+            log.info("Running cmd={}".format(cmd))
+            remote.run(
+                args=cmd,
+            )
+            time.sleep(10)
+    try:
+        yield
+
+    finally:
+        teuthology.stop_daemons_of_type(ctx, "osd", config['cluster'])
+
+def _fix_roles_map(ctx, config):
+    if not hasattr(ctx, 'managers'):
+        ctx.managers = {}
+    ctx.daemons.use_systemd=True
+    ready_cluster = ctx.cluster
+    cluster_name = config['cluster']
+    if not hasattr(ctx, 'new_remote_role'):
+        new_remote_role = dict()
+        ctx.new_remote_role = new_remote_role
+    else:
+        new_remote_role = ctx.new_remote_role
+    log.info("ROLES BEFORE:")
+    log.info(ctx.cluster.remotes)
+    firstmon = teuthology.get_first_mon(ctx, config, config['cluster'])
+    for remote, roles in ready_cluster.remotes.iteritems():
+        new_remote_role[remote] = []
+        generate_osd_list = True
+        for role in roles:
+            cluster, rol, id = teuthology.split_role(role)
+            if rol.startswith('osd'):
+                if generate_osd_list:
+                    # gather osd ids as seen on host
+                    out = StringIO()
+                    remote.run(args=[
+                                    'ps', '-eaf', run.Raw('|'), 'grep',
+                                    'ceph-osd', run.Raw('|'),
+                                    run.Raw('awk {\'print $13\'}')],
+                                stdout=out)
+                    osd_list_all = out.getvalue().split('\n')
+                    generate_osd_list = False
+                    osd_list = []
+                    for osd_id in osd_list_all:
+                        try:
+                            if type(int(osd_id)) is int:
+                                osd_list.append(osd_id)
+                        except ValueError:
+                            # ignore any empty lines as part of output
+                            pass
+                id = osd_list.pop()
+                log.info("Registering Daemon {rol} {id}".format(rol=rol, id=id))
+                ctx.daemons.add_daemon(remote, rol, id)
+                if len(role.split('.')) == 2:
+                    osd_role = "{rol}.{id}".format(rol=rol, id=id)
+                else:
+                    osd_role = "{c}.{rol}.{id}".format(c=cluster, rol=rol, id=id)
+                new_remote_role[remote].append(osd_role)
+            else:
+                new_remote_role[remote].append(role)
+
+    ctx.daemons.use_systemd=False
```

Comment:
I'm confused how this works since add_daemon()->register_daemon() has instantiated a bunch of SystemDState objects for the osds. Does this make the existing ctx.daemons users like CephManager work, or will that require follow-on work?

Longer term I assume we'll need further changes to make all the tests work with systemd-based daemons

Reply:
Yeah here osd will be using systemd but the rest we are not changing (mon, mds etc) so its a mixed system with systemd and non systemd components,  Some additional work is required later on to fix all thrash code to work in this scenario.